moveit_core/robot_state/include/moveit/robot_state/cartesian_interpolator.h

```diff
@@ -173,43 +171,63 @@ class CartesianInterpolator
                        std::vector<std::shared_ptr<RobotState>>& traj, const LinkModel* link,
                        const Eigen::Vector3d& direction, bool global_reference_frame, double distance,
                        const MaxEEFStep& max_step, const JumpThreshold& jump_threshold,
                        const GroupStateValidityCallbackFn& validCallback = GroupStateValidityCallbackFn(),
                        const kinematics::KinematicsQueryOptions& options = kinematics::KinematicsQueryOptions(),
-                       kinematics::KinematicsBase::IKCostFn cost_function = kinematics::KinematicsBase::IKCostFn());
+                       kinematics::KinematicsBase::IKCostFn cost_function = kinematics::KinematicsBase::IKCostFn())
+  {
+    return computeCartesianPath(start_state, group, traj, link, distance * direction, global_reference_frame, max_step,
+                                jump_threshold, validCallback, options, cost_function);
+  }
 
-  /** \brief Compute the sequence of joint values that correspond to a straight Cartesian path, for a particular group.
+  /** \brief Compute the sequence of joint values that correspond to a straight Cartesian path, for a particular frame.
 
      In contrast to the previous function, the Cartesian path is specified as a target frame to be reached (\e target)
-     for the origin of a robot link (\e link). The target frame is assumed to be either in a global reference frame or
-     in the local reference frame of the link. In the latter case (\e global_reference_frame is false) the \e target is
-     rotated accordingly. All other comments from the previous function apply. */
+     for a virtual frame attached to the robot \e link with the given \e link_offset.
+     The target frame is assumed to be specified either w.r.t. to the global reference frame or the virtual link frame.
+     This function returns the percentage (0..1) of the path that was achieved.
+     All other comments from the previous function apply. */
   static Percentage
   computeCartesianPath(RobotState* start_state, const JointModelGroup* group,
                        std::vector<std::shared_ptr<RobotState>>& traj, const LinkModel* link,
                        const Eigen::Isometry3d& target, bool global_reference_frame, const MaxEEFStep& max_step,
                        const JumpThreshold& jump_threshold,
                        const GroupStateValidityCallbackFn& validCallback = GroupStateValidityCallbackFn(),
                        const kinematics::KinematicsQueryOptions& options = kinematics::KinematicsQueryOptions(),
-                       kinematics::KinematicsBase::IKCostFn cost_function = kinematics::KinematicsBase::IKCostFn());
+                       kinematics::KinematicsBase::IKCostFn cost_function = kinematics::KinematicsBase::IKCostFn(),
+                       const Eigen::Isometry3d& link_offset = Eigen::Isometry3d::Identity());
```

Comment:
What's the advantage of `link_offset` compared to offsetting `target` accordingly? These functions already have way too many arguments, adding this will make them even more difficult to understand.

Reply:
There is a fundamental difference between adding the offset once to the target and considering the offset during the whole motion - but only when it comes to rotations. Consider rotating about this offset point (see the video [here](https://github.com/ros-planning/moveit_task_constructor/pull/380) for illustration):
In both approaches the _final_ end-effector poses will be identical. However, how you reach there is dramatically different:
- in the former case, you will perform a straigh-line motion along the secant of the circle
- in the latter case, you will perform a circular motion

Reply:
I get that. I just don't get why the Cartesian motion can't be computed outside of the IK+interpolation functions (at least that's what I thought these were), and then passed as `EigenSTL::vector_Isometry3d&` waypoints. To me it just looks like the functions do to much, and a separate function that transforms your waypoints before passing it to `computeCartesianPath()` would be more adequate.

Reply:
Such a "transformation" function would do exactly the same as `computeCartesianPath()`, just w/o IK computation. Hence, I deem it more reasonable to avoid the corresponding code duplication.